File: JS/quickSort.js
//quicksort implementation

let arr = [23, 11, 3, 2, 244, 133, 3433, 1, 2, 3, 100, 99];

function quickSort(start, stop) {
    if (start === stop) return;
    if (start > stop) return;

    let pivot = start;

    for (let i = start + 1; i <= stop; i++) {
        if (arr[i] < arr[pivot]) {
            //swap arrays XOR style
            arr[i] = arr[pivot] ^ arr[i];
            arr[pivot] = arr[i] ^ arr[pivot];
            arr[i] = arr[i] ^ arr[pivot];
            pivot = i;
        }
    }

    if (pivot < stop && start != pivot) quickSort(pivot, stop);
    if (start < pivot && start != pivot) quickSort(start, pivot);
    return 0;
}

quickSort(0, arr.length);
document.write(arr);

File: lab_practice/sort/quick_sort.py
def quick_sort(arr: list) -> list:
    return sort(arr,0,len(arr))

def sort(arr:list,start,stop):
    if start < stop:
        loc =  partition(arr,start,stop)
        sort(arr,0,loc-1)
        sort(arr,loc+1,stop)
    return arr


def partition(arr:list, start, stop):
    pivot = arr[stop-1]
    pivot_loc = start
    for i in range(start,stop):
        if arr[i] < pivot:
            arr[i],arr[pivot_loc] = arr[pivot_loc],arr[i]
            pivot_loc += 1

    arr[stop-1],arr[pivot_loc]= arr[pivot_loc],arr[stop-1]
    return pivot_loc
